### PR TITLE
Fix return_bias option in LayerNormLinear and LayerNormMLP

### PR DESCRIPTION
**Pull request by ptrendx**
# Description

return_bias option was ignored silently in LayerNormLinear and LayerNormMLP after the 2.0 refactor.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Change A
- Change B

# Checklist:

- [ ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by timmoon10 — APPROVED** [review]
Looks correct, especially as a quick bugfix.

`return_bias` is spaghetti code, so I'll summarize my understanding for future reference. Mcore performs some kernel fusions like bias-dropout-add and bias-add, which involve separating the bias compute from the linear GEMM. For an example, see this Transformer layer residual connection:
https://github.com/NVIDIA/Megatron-LM/blob/34013dbd37e94e17f1434d71b6ce9705e0acf6ca/megatron/core/transformer/transformer_layer.py#L451-L453
However, Mcore would like to stay consistent with PyTorch's convention of including the bias with the linear module (I presume for checkpointing). `return_bias` tells TE to not actually apply the bias, but to return it for use by Mcore. For some reason we still pass the bias into the autograd function, but then have a bunch of checks everywhere so that it isn't ever used in the forward or backward pass. In the future we should just pass `bias=None` into the autograd function if `return_bias=True`, so we can remove all the redundant logic with `use_bias=False`.

File: tests/pytorch/test_numerics.py [hunk 1]
```diff
@@ -344,7 +344,7 @@ def __init__(
         else:
             raise RuntimeError("Unsupported normalization")
 
-        self.linear = nn.Linear(in_features, out_features)
+        self.linear = nn.Linear(in_features, out_features, bias=not return_bias)
```
[the remark this answers is not in the document]
Comment (ptrendx) [reply tests/pytorch/test_numerics.py:347]: We would need to then apply it differently, since there is no option like return bias in the nn.Linear. So in the test I would need to not actually apply self.linear but rather functional linear with the weight from self.linear - do you want me to do that?

[the remark this answers is not in the document]
Comment (timmoon10) [reply tests/pytorch/test_numerics.py:347]: I think the right behavior is to apply the bias after the `te.Linear` to simulate Mcore doing bias-dropout-add.

[the remark this answers is not in the document]
Comment (ptrendx) [reply tests/pytorch/test_numerics.py:347]: Ah, ok, I think I see what you mean - so basically modify the TE path in the test to apply the returned bias. I can do that.

File: tests/pytorch/test_numerics.py [hunk 2]
```diff
@@ -1116,32 +1119,51 @@ def test_dpa_accuracy(dtype, bs, model):
...
+        super().__init__()
+        self.te_module = mod(**kwargs)
+        self.return_bias = kwargs["return_bias"]
+        self.bias = kwargs["bias"]
+
+    def forward(self, x):
+        if self.bias and self.return_bias:
+            out, bias = self.te_module(x)
+            return out + bias
+        return self.te_module(x)
```
Comment (timmoon10) [inline tests/pytorch/test_numerics.py:1133]: Do we handle the case where `bias=False` and `return_bias=True`?
```suggestion
        if self.return_bias:
            out, bias = self.te_module(x)
            if self.bias:
                out = out + bias
            return out
        return self.te_module(x)
```

**Review by timmoon10 — COMMENTED** [review]
Moving the `return_bias` out of the autograd function is a big improvement.

**Review by ksivaman — APPROVED** [review]
LGTM

**Comment by ptrendx** [comment]
The CI failures are the expected failures coming from paged attention with cuDNN 9.8. Merging.

**Comment by Marks101** [comment]
Hi @ptrendx and @ksivaman, could you maybe mention this in the known issues for version 2.1 and 2.0? We have been falling for this and it caused some trouble 🙈

**Comment by ptrendx** [comment]
Hi @Marks101, my apology about that :-(. I just added the note about it to release notes in 2.0/2.1 on GitHub.

**Comment by pggPL** [comment]
@ptrendx I noticed "/opt/transformerengine/tests/pytorch/test_numerics.py:1147: PytestCollectionWarning: cannot collect test class 'TestReturnBiasModule' because it has a __init__ constructor (from: tests/pytorch/test_numerics.py)" in the test logs.
